### ISSUE TITLE
unexpected generic case in LLVMNativeCallback (java.lang.IllegalStateException) in Nokogiri CI

### ISSUE THREAD

**Issue by flavorjones** [oracle/truffleruby#2203]
Hello there! I've been running Nokogiri's CI against both TR stable and TR nightly builds, and something changed in the past few weeks where the test suite's failure mode changed from behaviorial tests failing to a `java.lang.IllegalStateException` being raised and terminating the run.

---

See the logs at https://ci.nokogiri.org/teams/nokogiri-core/pipelines/nokogiri-pr/jobs/truffle-nightly/builds/65#L5fc58658:18 for the **last-known-OK** CI run (running `truffleruby 21.0.0-dev-c79719ba, like ruby 2.7.2, GraalVM CE Native [x86_64-linux]`):

<details>
  1) Failure:
TestXsltTransforms#test_0001_should not crash when given XPath 2.0 features [/tmp/build/4d9a0f57/nokogiri/test/test_xslt_transforms.rb:359]:
Expected /decimal/ to match # encoding: ASCII-8BIT
#    valid: true
"Generic errorGeneric errorGeneric error".

  2) Failure:
Nokogiri::XML::TestSaxEntityReference#test_sax_entity_reference_with_absolute_path [/tmp/build/4d9a0f57/nokogiri/test/xml/test_entity_reference.rb:196]:
Expected: ["Entity 'bar' not defined"]
  Actual: ["Warning."]

  3) Failure:
Nokogiri::XML::TestSaxEntityReference#test_more_sax_entity_reference [/tmp/build/4d9a0f57/nokogiri/test/xml/test_entity_reference.rb:209]:
Expected: ["Entity 'bar' not defined"]
  Actual: ["Warning."]

  4) Failure:
Nokogiri::XML::TestSaxEntityReference#test_more_sax_entity_reference_with_absolute_path [/tmp/build/4d9a0f57/nokogiri/test/xml/test_entity_reference.rb:209]:
Expected: ["Entity 'bar' not defined"]
  Actual: ["Warning."]

  5) Failure:
Nokogiri::XML::TestSaxEntityReference#test_sax_entity_reference [/tmp/build/4d9a0f57/nokogiri/test/xml/test_entity_reference.rb:196]:
Expected: ["Entity 'bar' not defined"]
  Actual: ["Warning."]
<summary>
1566 runs, 4408 assertions, 5 failures, 0 errors, 10 skips
</summary>
</details>

---

See the logs at https://ci.nokogiri.org/teams/nokogiri-core/pipelines/nokogiri-pr/jobs/truffle-nightly/builds/66#L5fc58999:35 for the **first-known-bad** CI run (running `truffleruby 21.0.0-dev-31d6113b, like ruby 2.7.2, GraalVM CE Native [x86_64-linux]`):

```
truffleruby: Unexpected internal exception in at_exit,
please report it to https://github.com/oracle/truffleruby/issues.

unexpected generic case in LLVMNativeCallback (java.lang.IllegalStateException)
	from com.oracle.truffle.llvm.runtime.interop.nfi.LLVMNativeWrapper$CallbackHelperNode.doGeneric(LLVMNativeWrapper.java:122)
	from com.oracle.truffle.llvm.runtime.interop.nfi.LLVMNativeWrapperFactory$CallbackHelperNodeGen.executeAndSpecialize(LLVMNativeWrapperFactory.java:119)
	from com.oracle.truffle.llvm.runtime.interop.nfi.LLVMNativeWrapperFactory$CallbackHelperNodeGen.execute(LLVMNativeWrapperFactory.java:58)
	from com.oracle.truffle.llvm.runtime.interop.nfi.LLVMNativeWrapper.execute(LLVMNativeWrapper.java:84)
	from com.oracle.truffle.llvm.runtime.interop.nfi.LLVMNativeWrapperGen$InteropLibraryExports$Cached.execute(LLVMNativeWrapperGen.java:81)
	from com.oracle.truffle.nfi.CallSignatureNode$OptimizedCallClosureNode.doCall(CallSignatureNode.java:218)
	from com.oracle.truffle.nfi.CallSignatureNodeFactory$OptimizedCallClosureNodeGen.execute(CallSignatureNodeFactory.java:473)
	from com.oracle.truffle.nfi.CallSignatureNode$CallSignatureRootNode.execute(CallSignatureNode.java:238)
	from org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:591)
/tmp/build/4d9a0f57/nokogiri/ext/nokogiri/html_sax_push_parser.c:23:in `native_write'
	from /usr/local/lib/truffle/truffle/cext_ruby.rb:39:in `native_write'
	from /tmp/build/4d9a0f57/nokogiri/lib/nokogiri/html/sax/push_parser.rb:24:in `write'
	from /tmp/build/4d9a0f57/nokogiri/test/html/sax/test_push_parser.rb:78:in `test_broken_encoding'
	from /usr/local/bundle/gems/minitest-5.14.2/lib/minitest/test.rb:98:in `block (3 levels) in run'
```

---

On nokogiri's master branch, the versions of last-known-good and first-known-bad truffleruby are the same. See https://ci.nokogiri.org/teams/nokogiri-core/pipelines/nokogiri/jobs/truffle-nightly/builds/28 and https://ci.nokogiri.org/teams/nokogiri-core/pipelines/nokogiri/jobs/truffle-nightly/builds/29 for more details.

**Comment by eregon** [comment]
Thank you for the detailed report, this is a Sulong or NFI bug, I've filed GR-28462 internally to track this.

**Comment by eregon** [comment]
@flavorjones The logs above are no longer available.
I guess the main commands to reproduce are (looking at https://github.com/sparklemotion/nokogiri/blob/master/.github/workflows/macos.yml):
```
clone && cd
bundle install
bundle exec rake compile
bundle exec rake test
```
right?

**Comment by rschatz** [comment]
Should be fixed on latest master of oracle/graal (https://github.com/oracle/graal/commit/a2aced5ed06803f278ae2994976c0294a16d76fc). Backport to the 21.0 release branch should also land soon.

This still needs an import update in the oracle/truffleruby repo. I assume this is going to happen soon anyway (@eregon  should we do an extra update for this fix?).

**Comment by eregon** [comment]
Yes, I'll update the graal import in TruffleRuby soon, and then I'll close this, so that nightly builds also pick up the fix.

**Comment by eregon** [comment]
Fixed in 21.0.0 and on master.
